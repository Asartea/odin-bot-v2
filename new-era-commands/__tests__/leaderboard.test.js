const { execute } = require('../slash/leaderboard');
const axios = require('axios');
const { generateLeaderData } = require('../../botCommands/mockData');

/* eslint-disable */
/* eslint max-classes-per-file: ["error", 2] */

class GuildMembersMock {
  members;

  cache;

  constructor(users) {
    this.members = users;
    this.cache = {
      get: (id) => users.filter((member) => member.discord_id === id)[0],
    };
  }
}

class GuildMock {
  members;

  member;

  constructor(users) {
    this.members = new GuildMembersMock(users);
    this.member = (user) => users.filter((member) => member === user)[0];
  }
}

describe('Reply the correct information for Server leaderboard', () => {
  let limit = null;
  let offset = null;
  const interactionMock = {
    options: {
      getSubcommand: () => 'ranking',
      getInteger: (string) => {
        if (string === 'limit') {
          return limit;
        } else if (string === "offset") {
          return offset;
        }
      }
    },
    reply: jest.fn((message) => expect(message).toMatchSnapshot())
  }

  afterEach(() => axios.get.mockReset());

  const setUpAxiosMock = (data) => axios.get = jest.fn().mockResolvedValue({ data });

  it.skip('Returns correct output if provided no options', () => {
    const members = generateLeaderData(5);
    interactionMock.guild = new GuildMock(members);
    setUpAxiosMock(members)

    execute(interactionMock);
  });

  it.skip('Returns limits results to 25 if limit not provided', () => {
    const members = generateLeaderData(30);
    interactionMock.guild = new GuildMock(members);
    setUpAxiosMock(members);

    execute(interactionMock);
  });

  it.skip("Returned users matches limit", () => {
    limit = 8;
    const members = generateLeaderData(30);
    interactionMock.guild = new GuildMock(members);
    setUpAxiosMock(members);

    execute(interactionMock);
  })

  it.skip("Limit defaulst to 25 used if limit provided lower than 1", () => {
    limit = 0;
    const members = generateLeaderData(5);
    interactionMock.guild = new GuildMock(members);
    setUpAxiosMock(members);
    execute(interactionMock);
  })

  it.skip("Limit defaults to 25 when limit provided is higher than 25", () => {
    limit = 50;
    const members = generateLeaderData(70);
    interactionMock.guild = new GuildMock(members);
    setUpAxiosMock(members);

    execute(interactionMock);
  })

  it.skip("Retruns users starting from offset", () => {
    offset = 5;
    limit = 5;
    const members = generateLeaderData(50);
    interactionMock.guild = new GuildMock(members);
    setUpAxiosMock(members);

    execute(interactionMock);
  })

  //       expect(
  //         await commands.leaderboard.cb({
  //           guild: new GuildMock(members),
  //           content: '!leaderboard n=2 start=3',
  //         }),
  //       ).toMatchSnapshot();
  //       expect(
  //         await commands.leaderboard.cb({
  //           guild: new GuildMock(members),
  //           content: '!leaderboard start=3',
  //         }),
  //       ).toMatchSnapshot();
  //       expect(
  //         await commands.leaderboard.cb({
  //           guild: new GuildMock(members),
  //           content: '!leaderboard n=2',
  //         }),
  //       ).toMatchSnapshot();
  //       expect(
  //         await commands.leaderboard.cb({
  //           guild: new GuildMock(members),
  //           content: '!leaderboard n=2 start=wtf',
  //         }),
  //       ).toMatchSnapshot();
  //       expect(
  //         await commands.leaderboard.cb({
  //           guild: new GuildMock(members),
  //           content: '!leaderboard n=wtf start=3',
  //         }),
  //       ).toMatchSnapshot();
  //       expect(
  //         await commands.leaderboard.cb({
  //           guild: new GuildMock(members),
  //           content: '!leaderboard n=25 start=9999',
  //         }),
  //       ).toMatchSnapshot();
  //     });
  //   });
  // });

  // describe('!points', () => {
  //   const author = {
  //     id: '111333',
  //     displayName: 'odin'
  //   }
  //   const mentionedUser = {
  //     id: '222444',
  //     displayName: "NotOdin"
  //   };
  //
  //   const data = {
  //     author,
  //     guild: Guild([author, mentionedUser])
  //   }
  //
  //   describe('regex', () => {
  //     it.each([
  //       ['!points <@!123456789>'],
  //       ['let me check out my !points <@!123456789>'],
  //       ['!points <@!123456789> <@!123456789>-v2'],
  //       ['!points'],
  //     ])('correct strings trigger the callback', (string) => {
  //       expect(commands.points.regex.test(string)).toBeTruthy();
  //     });
  //
  //     it('returns author points information if no other user were provided', async () => {
  //       data.content = '!points';
  //       const axiosData = {
  //         data: {
  //           points: 5,
  //           rank: 1,
  //         }
  //       };
  //
  //       axios.get = jest.fn(() => axiosData);
  //       const reply = await commands.points.cb(data);
  //
  //       expect(reply).toMatchSnapshot();
  //       expect(axios.get).toHaveBeenCalled();
  //     });
  //
  //     it('return correct user points information if specified', async () => {
  //       data.content = '!points <@222444>';
  //       const axiosData = {
  //         data: {
  //           points: 20,
  //           rank: 1,
  //         }
  //       };
  //
  //       axios.get = jest.fn(() => axiosData);
  //       const reply = await commands.points.cb(data);
  //
  //       expect(reply).toMatchSnapshot();
  //       expect(axios.get).toHaveBeenCalled();
  //     });
  //
  //     it('returns correct msg when user has no points', async () => {
  //       data.content = '!points <@222444>';
  //       const axiosData = {
  //         data: {
  //           message: "unable to find that user",
  //         }
  //       };
  //
  //       axios.get = jest.fn(() => axiosData);
  //
  //       const reply = await commands.points.cb(data);
  //       expect(reply).toMatchSnapshot();
  //       expect(axios.get).toHaveBeenCalled();
  //     });
  //
  //     it('GET request not called if user not on disord', async () => {
  //       data.content = '!points <@11111111>';
  //
  //       axios.get = jest.fn();
  //       const reply = await commands.points.cb(data);
  //
  //       expect(reply).toMatchSnapshot();
  //       expect(axios.get).not.toHaveBeenCalled();
  //     });
  //
  //     it('format the points word properly for 1 point', async () => {
  //       data.content = '!points';
  //       const axiosData = {
  //         data: {
  //           points: 1,
  //           rank: 1,
  //         }
  //       }
  //
  //       axios.get = jest.fn(() => axiosData);
  //       const reply = await commands.points.cb(data);
  //
  //       expect(reply).toMatchSnapshot();
  //       expect(axios.get).toHaveBeenCalled();
  //     });
  //
  //     it('show correct user rank', async () => {
  //       data.content = '!points <@222444>';
  //       const axiosData = {
  //         data: {
  //           points: 20,
  //           rank: 50
  //         }
  //       };
  //
  //       axios.get = jest.fn(() => axiosData);
  //       const reply = await commands.points.cb(data);
  //
  //       expect(reply).toMatchSnapshot();
  //       expect(axios.get).toHaveBeenCalled();
  //     })
  //   });
  // });
});
