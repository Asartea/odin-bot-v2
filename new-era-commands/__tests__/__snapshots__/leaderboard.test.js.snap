// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ranking subcommand Limit defaults to 25 if invalid characters provided 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Server Ranking",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
6 - user105 [995 points] 
7 - user106 [994 points] 
8 - user107 [993 points] 
9 - user108 [992 points] 
10 - user109 [991 points] 
11 - user110 [990 points] 
12 - user111 [989 points] 
13 - user112 [988 points] 
14 - user113 [987 points] 
15 - user114 [986 points] 
16 - user115 [985 points] 
17 - user116 [984 points] 
18 - user117 [983 points] 
19 - user118 [982 points] 
20 - user119 [981 points] 
21 - user120 [980 points] 
22 - user121 [979 points] 
23 - user122 [978 points] 
24 - user123 [977 points] 
25 - user124 [976 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`ranking subcommand Limit defaults to 25 if limit not provided 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Server Ranking",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
6 - user105 [995 points] 
7 - user106 [994 points] 
8 - user107 [993 points] 
9 - user108 [992 points] 
10 - user109 [991 points] 
11 - user110 [990 points] 
12 - user111 [989 points] 
13 - user112 [988 points] 
14 - user113 [987 points] 
15 - user114 [986 points] 
16 - user115 [985 points] 
17 - user116 [984 points] 
18 - user117 [983 points] 
19 - user118 [982 points] 
20 - user119 [981 points] 
21 - user120 [980 points] 
22 - user121 [979 points] 
23 - user122 [978 points] 
24 - user123 [977 points] 
25 - user124 [976 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`ranking subcommand Limit defaults to 25 if limit provided is higher than 25 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Server Ranking",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
6 - user105 [995 points] 
7 - user106 [994 points] 
8 - user107 [993 points] 
9 - user108 [992 points] 
10 - user109 [991 points] 
11 - user110 [990 points] 
12 - user111 [989 points] 
13 - user112 [988 points] 
14 - user113 [987 points] 
15 - user114 [986 points] 
16 - user115 [985 points] 
17 - user116 [984 points] 
18 - user117 [983 points] 
19 - user118 [982 points] 
20 - user119 [981 points] 
21 - user120 [980 points] 
22 - user121 [979 points] 
23 - user122 [978 points] 
24 - user123 [977 points] 
25 - user124 [976 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`ranking subcommand Limit defaults to 25 if limit provided lower than 1 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Server Ranking",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
6 - user105 [995 points] 
7 - user106 [994 points] 
8 - user107 [993 points] 
9 - user108 [992 points] 
10 - user109 [991 points] 
11 - user110 [990 points] 
12 - user111 [989 points] 
13 - user112 [988 points] 
14 - user113 [987 points] 
15 - user114 [986 points] 
16 - user115 [985 points] 
17 - user116 [984 points] 
18 - user117 [983 points] 
19 - user118 [982 points] 
20 - user119 [981 points] 
21 - user120 [980 points] 
22 - user121 [979 points] 
23 - user122 [978 points] 
24 - user123 [977 points] 
25 - user124 [976 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`ranking subcommand Offset cannot exceed max users length 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Server Ranking",
          "value": "26 - user125 [975 points] 
27 - user126 [974 points] 
28 - user127 [973 points] 
29 - user128 [972 points] 
30 - user129 [971 points] 
31 - user130 [970 points] 
32 - user131 [969 points] 
33 - user132 [968 points] 
34 - user133 [967 points] 
35 - user134 [966 points] 
36 - user135 [965 points] 
37 - user136 [964 points] 
38 - user137 [963 points] 
39 - user138 [962 points] 
40 - user139 [961 points] 
41 - user140 [960 points] 
42 - user141 [959 points] 
43 - user142 [958 points] 
44 - user143 [957 points] 
45 - user144 [956 points] 
46 - user145 [955 points] 
47 - user146 [954 points] 
48 - user147 [953 points] 
49 - user148 [952 points] 
50 - user149 [951 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`ranking subcommand Offset defaults to 0 if invalid characters provided 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Server Ranking",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
6 - user105 [995 points] 
7 - user106 [994 points] 
8 - user107 [993 points] 
9 - user108 [992 points] 
10 - user109 [991 points] 
11 - user110 [990 points] 
12 - user111 [989 points] 
13 - user112 [988 points] 
14 - user113 [987 points] 
15 - user114 [986 points] 
16 - user115 [985 points] 
17 - user116 [984 points] 
18 - user117 [983 points] 
19 - user118 [982 points] 
20 - user119 [981 points] 
21 - user120 [980 points] 
22 - user121 [979 points] 
23 - user122 [978 points] 
24 - user123 [977 points] 
25 - user124 [976 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`ranking subcommand Offset defaults to 0 if negative value provided 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Server Ranking",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
6 - user105 [995 points] 
7 - user106 [994 points] 
8 - user107 [993 points] 
9 - user108 [992 points] 
10 - user109 [991 points] 
11 - user110 [990 points] 
12 - user111 [989 points] 
13 - user112 [988 points] 
14 - user113 [987 points] 
15 - user114 [986 points] 
16 - user115 [985 points] 
17 - user116 [984 points] 
18 - user117 [983 points] 
19 - user118 [982 points] 
20 - user119 [981 points] 
21 - user120 [980 points] 
22 - user121 [979 points] 
23 - user122 [978 points] 
24 - user123 [977 points] 
25 - user124 [976 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`ranking subcommand Offset defaults to 0 if not provided 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Server Ranking",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
6 - user105 [995 points] 
7 - user106 [994 points] 
8 - user107 [993 points] 
9 - user108 [992 points] 
10 - user109 [991 points] 
11 - user110 [990 points] 
12 - user111 [989 points] 
13 - user112 [988 points] 
14 - user113 [987 points] 
15 - user114 [986 points] 
16 - user115 [985 points] 
17 - user116 [984 points] 
18 - user117 [983 points] 
19 - user118 [982 points] 
20 - user119 [981 points] 
21 - user120 [980 points] 
22 - user121 [979 points] 
23 - user122 [978 points] 
24 - user123 [977 points] 
25 - user124 [976 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`ranking subcommand Offset show the last users depending on limit if offset is too high 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Server Ranking",
          "value": "46 - user145 [955 points] 
47 - user146 [954 points] 
48 - user147 [953 points] 
49 - user148 [952 points] 
50 - user149 [951 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`ranking subcommand Return the correct limit, moving offset backward if necessary 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Server Ranking",
          "value": "51 - user150 [950 points] 
52 - user151 [949 points] 
53 - user152 [948 points] 
54 - user153 [947 points] 
55 - user154 [946 points] 
56 - user155 [945 points] 
57 - user156 [944 points] 
58 - user157 [943 points] 
59 - user158 [942 points] 
60 - user159 [941 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`ranking subcommand Returned users length matches limit 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Server Ranking",
          "value": "1 - user100 [1000 points] :tada: 
2 - user101 [999 points] 
3 - user102 [998 points] 
4 - user103 [997 points] 
5 - user104 [996 points] 
6 - user105 [995 points] 
7 - user106 [994 points] 
8 - user107 [993 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`ranking subcommand Returns users starting from offset 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "Server Ranking",
          "value": "6 - user105 [995 points] 
7 - user106 [994 points] 
8 - user107 [993 points] 
9 - user108 [992 points] 
10 - user109 [991 points] 
11 - user110 [990 points] 
12 - user111 [989 points] 
13 - user112 [988 points] 
14 - user113 [987 points] 
15 - user114 [986 points] 
16 - user115 [985 points] 
17 - user116 [984 points] 
18 - user117 [983 points] 
19 - user118 [982 points] 
20 - user119 [981 points] 
21 - user120 [980 points] 
22 - user121 [979 points] 
23 - user122 [978 points] 
24 - user123 [977 points] 
25 - user124 [976 points] 
26 - user125 [975 points] 
27 - user126 [974 points] 
28 - user127 [973 points] 
29 - user128 [972 points] 
30 - user129 [971 points] 
",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`user subcommand Format the points word properly for 1 point 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "User Rank",
          "value": "Someone has 1 point!",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`user subcommand Formats the points word properly for 0 or more than 1 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "User Rank",
          "value": "Dog has 20 points!",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`user subcommand Formats the points word properly for 0 or more than 1 2`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "User Rank",
          "value": "Dog has 0 points!",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`user subcommand Return correct rank if user rank recieved 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "User Rank",
          "value": "5 - Cat has 20 points!",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;

exports[`user subcommand Return correct reply when user has no points 1`] = `
{
  "embeds": [
    {
      "color": 13407555,
      "fields": [
        {
          "name": "User Rank",
          "value": "NotOdin has 0 points!",
        },
      ],
      "title": "Leaderboard",
    },
  ],
}
`;
